### PR TITLE
Fix tests on windows, enable windows to github auto-testing

### LINKED ISSUE #39: Make sure tests pass on Windows, add windows to GitHub Actions workflow
Currently, the `test_evolutionary_agent` test uses *nix specific syntax in the `virtualenv.run()` command, which is thus failing:
```
>       virtualenv.run(f"cd {os.getcwd()}/example_agents/evolutionary_agent; "
                       "ls -al; "
                       "pip install -r requirements.txt; "
                       "agentos run --max-iters 5 agent.py "
                       "gym.envs.classic_control.CartPoleEnv")
```

These commands should use `os.sep`, `os.pardir`, etc. to be OS agnostic.

### PR DESCRIPTION
**Pull request by andyk**
* Fixes `test_evolutionary_agent` pytest that was failing on Windows because it was using bash specific path syntax.
* Updates Github Actions workflow so that
  * tests automatically run on windows (in addition to existing envs that include MacOS and Linux)
  * tests automatically run after a push to any branch (but only on PRs to master, which is current behavior). This will make sure the tests are run when we push changes to our forks so that we can see test github test results before we submit our PRs. **NOTE** that creating a draft PR is another way to run the tests while preparing a PR.
* Updates `example_agents/predictive_coding/free_energy_tutorial` to follow conventions for agents (by making the Env it defines extend `gym.Env`)
* improves docstrings on some of core API
  
Fixes #39.

### CODE REVIEW

**Comment by andyk** [comment]
@nickjalbert, I believe this PR that fixes/adds windows testing is ready for review!

**Review by nickjalbert — APPROVED** [review]
You'll have to show me how you set up your windows env so I can create one as well (or is it already in the docs?).  

Also, I notice you added the comment to some of the requirements files:  

```
# These dependencies are also listed in the <proj_root>/dev-requirements.txt
# If you change these, make the same changes there.
```

This makes me worry we'll forget to update something at some point!  

I'm guessing you're trying to split apart the requirements for the example agents in case some big and bulky agent with a bunch of dependencies gets built.  I wonder, though, if it'd make sense to roll all the requirements up into a single file and force heavyweight agents to be in their own repo and use agentos as a library (instead of including them as example agents).

Anyway, LGTM; merge away!  But let's talk about the requirements strategy offline when we get a chance.

File: example_agents/rl_agents/requirements.txt [hunk 1]
```diff
@@ -1,4 +1,7 @@
+# These dependencies are also listed in the <proj_root>/dev-requirements.txt
+# If you change these, make the same changes there.
+
 -e ../../
-tensorflow==2.4.0  # reinforce, dqn
```
[the remark this answers is not in the document]
Comment (andyk) [reply example_agents/rl_agents/requirements.txt:2]: Yeah I tried 2.4.0 and it seemed to break at least one test. I didn’t dig into it but I did a quick google and it looks like 2.2.2 at least fixes a security vulnerability that is in 2.2.0.

File: example_agents/chatbot/conda_env.yaml [hunk 2]
```diff
@@ -5,4 +5,4 @@ dependencies:
   - pip
   - pip:
     - -e ../../
-    - numpy
+    - numpy==1.18.5
```
[the remark this answers is not in the document]
Comment (andyk) [reply example_agents/chatbot/conda_env.yaml:8]: yeah conda is another virtual env and dependency management system similar in spirit to pipenv. MLflow uses conda, and currently the AgentOS CLI has a light integration with MLflow.

Only 2 of the example agents use Conda currently, and one of them (the agent that wraps rllib) is currently commented out.

File: dev-requirements.txt [hunk 3]
```diff
@@ -16,5 +16,7 @@ flake8==3.8.3
 matplotlib==3.2.1
 pytest==3.2.1
 pytest-virtualenv==1.7.0
-tensorflow==2.2.0  # reinforce, dqn
-tensorflow-probability==0.10.0  # for example_agents/rl_agents/reinforce_agent.py
+tensorflow==2.2.2  # reinforce, dqn, evolutionary_agent
```
[the remark this answers is not in the document]
Comment (andyk) [reply dev-requirements.txt:19]: See below

File: example_agents/rl_agents/requirements.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (nickjalbert) [reply example_agents/rl_agents/requirements.txt:2]: Makes sense and at least dependabot won't yell at us anymore! 😄   

Should we file an issue to upgrade to 2.4.0 so it stays on the radar?

**Comment by andyk** [comment]
Yeah, I agree that having the dependencies for example agents duplicated is asking for trouble.

I originally had intended I have each example agent be as standalone as possible,so that folks can get the best sense of how an Agent Directory is intended to encapsulate everything about an agent, including its dependencies.

I also wanted to demonstrate how different agents can make their own choice for dependency management tooling, so I wanted to show some using conda, and some using virtualenv (and eventually some using pipenv).

But that slightly complicates the centralized agentos testing strategy, since then you need to create a separate env for each example Agent, then install agentos and that agents dependencies in that env, and then run that agents tests in it. That’s what I was doing originally, but it got cumbersome enough that I broke down and started consolidating dependencies for example agents into the [what was then called] test-requirements.txt.

Setting up a virtual env and installing dependencies from scratch (esp. tensorflow, which is heavyweight) for each test also made the tests take longer to run.

We should probably come up with a well thought out set of conventions for example agents.

It should also address how we decide what type of agent goes into the example_agents dir vs into <proj_root>/agentos/agents.py which currently only contains RandomAgent.

It seems like we will want be able to build and test the agentos core functionality—including APIs and tooling (e.g., CLI)... basically anything in the agentos dir—independent of example agents, which are likely to be heavier weight, and have much more extensive dependencies.

Using an entirely different repo (or many repos) for example agents is a good option, but that also makes them more tricky to maintain and more, less easy for new users to try out or explore, and more likely that we neglect to update them as the core expands.

We could just keep them in the same repo like they are now, but separate out there dependencies and build file from the core stuff.

However, having 3 places agents can live (agents.py, example_agents dir, and separate repos) is probably too confusing and complex at this stage of the project.

**Comment by andyk** [comment]
I created #44 to track creation of a design doc for improved example agent processes, conventions, tooling.

File: test_all.py [hunk 4]
```diff
@@ -62,51 +70,80 @@ def test_chatbot(capsys):
...
     # TODO(andyk): also test CommandLineListener
 
 
-def test_rl_agents():
-    from agentos import run_agent
-    from example_agents.rl_agents.reinforce_agent import ReinforceAgent
-    from gym.envs.classic_control import CartPoleEnv
-
-    run_agent(ReinforceAgent, CartPoleEnv, max_iters=10)
+def run_agent_in_dir(
```
Comment (nickjalbert) [inline test_all.py:91]: Ah cool!  so pytest-virtualenv gives you a place to install your requirements during the test and it looks like the example agents all have their own independent requirements.  Nice way to eliminate the duplication!

What does the `-e` in the example agent requirement files do? Install the `agentos` library for the test?

Comment (andyk) [reply test_all.py:91]: Yep, you got it.

And yes, `-e` installs a package from source (short for editable).

This way the example agents can run against the version of agentos that is in the project folder on which they are located, and thus actually exercise any code that a user might have changed/added.

**Pull request merged**